### PR TITLE
Include `RuntimeLong.sjsir` as a resource of the linker.

### PR DESCRIPTION
**Pull request by sjrd**
`RuntimeLong` is basically an implementation of the linker. It shouldn't be part of the standard library. In the past, this has caused troubles when we wanted to improve the encoding of longs, as we had to keep backward and forward binary compatibility between versions of `RuntimeLong`.

In this commit, we move `RuntimeLong.scala` in a separate project, which is not published. Instead, its compiled products are stored as resources of the linker. The fact that we have 3 variants of the linker complicates things a bit:

* For the published JVM artifact, we use genuine resources.
* For the published JS artifact, we use source generation to embed the resources as constants.
* For the build itself, we cannot do either of those things, so instead we manually inject the products of `RuntimeLong.scala` on the `fullClasspath` of our projects, to enable linking.

The last bullet is concerning wrt. testing, as most of our tests happen in the build itself, and we want to minimize environmental differences between the build and the published linker. This issue is somewhat addressed by the unit tests of the linker and the smoke tests in sbt-plugin-test.

### CODE REVIEW

File: project/Build.scala [hunk 1]
```diff
@@ -721,14 +749,108 @@ object Build {
...
           parallelCollectionsDependencies(scalaVersion.value)
       ),
+
+      resourceGenerators in Compile += Def.task {
+        val baseResourceDir = (resourceManaged in Compile).value
+        val resourceDir = baseResourceDir / "org/scalajs/linker/backend/emitter"
+
+        /* Awful hack to execute `products in (linkerPrivateLibrary, Compile)`
+         * as if we were in ++2.12.8, regardless of the current version of the
+         * linker and other projects.
```
[the remark this answers is not in the document]
Comment (olafurpg) [reply project/Build.scala]: Glad it worked!

**Comment by gzm0** [comment]
Question, would the following work:

- Define a well known location for the results of `linkerPrivateLib` (`<root>/target/linkerPrivateLibResources`)
- Add that directory to the runtime classpath of the build (so in `project/build.sbt`).
- In MyScalaJSPlugin, make `fastOptJS` and `fullOptJS` depend on a task that copies the private linker library to that path.

If that works, we could get rid of the `== null` check and load things always the same way when on JVM.

**Comment by sjrd** [comment]
Unfortunately, it does not work. :( I made a branch here with an attempt:
https://github.com/scala-js/scala-js/compare/master...sjrd:rtlong-in-linker-resource-2

This will find the resource if we first make sure they exist, then *reload* (or quit sbt and relaunch), and only then try to access the resource. But it won't find the resource right after having created it. I guess the classloader only identifies resources when it is loaded, and one cannot add files after the fact and expect the classloader to be able to load them.

File: project/Build.scala [hunk 2]
```diff
@@ -721,14 +749,108 @@ object Build {
...
+
   lazy val linkerJS: Project = (project in file("linker/js")).enablePlugins(
       MyScalaJSPlugin
   ).settings(
       commonLinkerSettings,
       crossVersion := ScalaJSCrossVersion.binary,
+
+      sourceGenerators in Compile += Def.task {
+        val dir = (sourceManaged in Compile).value
+        val privateLibProducts = (products in (linkerPrivateLibrary, Compile)).value
```
Comment (gzm0) [inline project/Build.scala]: Shouldn't this also take the 2.12.8 RuntimeLong for consistency at least?

**Comment by gzm0** [comment]
Ugh... Thanks for trying. So basically this means we are stuck with the build specific check in production code or we use a class loader to load the linker in the Build which would require horrible changes to `ScalaJSPluginInternal` :(

So it seems we just have to live with this then :S

File: project/Build.scala (hunk 2 above)
Comment (sjrd) [reply project/Build.scala]: Yes, indeed. That would be better.

**Comment by sjrd** [comment]
> or we use a class loader to load the linker in the Build which would require horrible changes to ScalaJSPluginInternal

We might want to do that for other reasons, so maybe at some point it's going to become desirable. Things that are already pushing for something like that:

* It would allow us to drop 2.10 support (and 2.11 while we're at it) for the `linker`. That's desirable because there are a number of pattern matches where we had to work around 2.10 compiler crashes.
* It would decouple the binary API of the linker from that of the sbt plugin. The latter could then more easily be guaranteed to stay stable forever.
* It would force us to feel the "pain" of having a build tool that can't just integrate the data structures of its linker, so maybe we would be better prompted to design the interface in a way that is friendlier to other build tools (who for the most part do not have this luxury).

It's just something I've had on my mind for the past couple months, since it was brought to my attention that other build tools all have to class-load our linker, rather than just call into it.

**Comment by gzm0** [comment]
Hmmm... I don't really understand how using class loading changes anything w.r.t. binary compatibility. But yes, using the API in a similar way than most (all?) other build tools is definitely a desirable thing.

So do you think we should do this first or live with the `== null`?

**Comment by gzm0** [comment]
SG

File: project/Build.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (dwijnand) [reply project/Build.scala]: Out of task graph execution is dangerous.  See references to `update` in sbt-buildinfo.

**Comment by sjrd** [comment]
Superseded by #3817.

**Pull request closed without merging**